File: api/Net.go
// Package api 对应 Python 版 huawei_lte_api.api 的全部 API 分组。
package api

import (
	"fmt"

	"github.com/lvcdy/huawei-lte-api-go/session"
)

// Net 对应 Net.py。
type Net struct {
	*session.ApiGroup
}

// NewNet 创建 Net API 分组。
func NewNet(s *session.Session) *Net {
	return &Net{ApiGroup: session.NewApiGroup(s)}
}

// CurrentPlmn 获取当前 PLMN。对应 net/current-plmn。
func (n *Net) CurrentPlmn() (map[string]interface{}, error) {
	return n.S.Get("net/current-plmn", nil, "api")
}

// NetMode 获取当前网络模式。对应 net/net-mode。
func (n *Net) NetMode() (map[string]interface{}, error) {
	return n.S.Get("net/net-mode", nil, "api")
}

// SetNetMode 设置网络模式。对应 net/net-mode (post_set)。
//
// lteband/networkband 为 int 时自动转小写十六进制（如 0x80005 → "80005"）；
// 为 string 时原样使用。networkmode 为 string 时原样使用。
func (n *Net) SetNetMode(lteband interface{}, networkband interface{}, networkmode interface{}) (interface{}, error) {
	return n.S.PostSet("net/net-mode", session.O(
		"NetworkMode", networkModeValue(networkmode),
		"NetworkBand", bandHexValue(networkband),
		"LTEBand", bandHexValue(lteband),
	), false, "api", false, false)
}

// networkModeValue 返回 networkmode 的字符串值。
// NetworkMode 为 string 型枚举，其底层是 string，经 fmt.Sprintf("%v") 即可恢复。
func networkModeValue(v interface{}) string {
	if s, ok := v.(string); ok {
		return s
	}
	return fmt.Sprintf("%v", v)
}

// bandHexValue 返回 band 的小写十六进制字符串（无 0x 前缀）。
// string 原样返回。
func bandHexValue(v interface{}) string {
	if s, ok := v.(string); ok {
		return s
	}
	return fmt.Sprintf("%x", v)
}

// Network 获取网络信息。对应 net/network。
func (n *Net) Network() (map[string]interface{}, error) {
	return n.S.Get("net/network", nil, "api")
}

// SetNetwork 设置网络模式与频段。对应 net/network (post_set)。
func (n *Net) SetNetwork(networkmode string, networkband string) (interface{}, error) {
	return n.S.PostSet("net/network", session.O(
		"NetworkMode", networkmode,
		"NetworkBand", networkband,
	), false, "api", false, false)
}

// Register 获取网络注册状态。对应 net/register。
func (n *Net) Register() (map[string]interface{}, error) {
	return n.S.Get("net/register", nil, "api")
}

// SetRegister 设置网络注册模式。对应 net/register (post_set)。
func (n *Net) SetRegister(mode string, plmn string, rat string) (interface{}, error) {
	return n.S.PostSet("net/register", session.O(
		"Mode", mode,
		"Plmn", plmn,
		"Rat", rat,
	), false, "api", false, false)
}

// NetModeList 获取可用网络模式列表。对应 net/net-mode-list。
func (n *Net) NetModeList() (map[string]interface{}, error) {
	return n.S.Get("net/net-mode-list", nil, "api")
}

// PlmnList 获取可用 PLMN 列表。对应 net/plmn-list。
func (n *Net) PlmnList() (map[string]interface{}, error) {
	return n.S.Get("net/plmn-list", nil, "api")
}

// NetFeatureSwitch 获取网络功能开关状态。对应 net/net-feature-switch。
func (n *Net) NetFeatureSwitch() (map[string]interface{}, error) {
	return n.S.Get("net/net-feature-switch", nil, "api")
}

// CellInfo 获取小区信息。对应 net/cell-info。
func (n *Net) CellInfo() (map[string]interface{}, error) {
	return n.S.Get("net/cell-info", nil, "api")
}

// CspsState 获取 CSPS 状态。对应 net/csps_state。
func (n *Net) CspsState() (map[string]interface{}, error) {
	return n.S.Get("net/csps_state", nil, "api")
}

// Reconnect 重新连接网络。对应 net/reconnect (post_set)。
func (n *Net) Reconnect() (interface{}, error) {
	return n.S.PostSet("net/reconnect", session.O("ReconnectAction", 1), false, "api", false, false)
}

File: src/api/net.rs
//! Net API group (`api/Net.py`).

use serde_json::Value;

use crate::connection::Connection;
use crate::enums::net::NetworkMode;
use crate::errors::Result;
use crate::tools::map_of;
use crate::xml::XmlMap;

use super::{get_value, post_set};

/// Net API group.
pub struct Net<'a> {
    conn: &'a Connection,
}

impl<'a> Net<'a> {
    /// Build a group bound to `conn`.
    pub fn new(conn: &'a Connection) -> Self {
        Net { conn }
    }

    /// `net/current-plmn`. Current Public Land Mobile Network.
    pub fn current_plmn(&self) -> Result<Value> {
        get_value(self.conn, "api/net/current-plmn")
    }

    /// `net/net-mode`. Current network mode.
    pub fn net_mode(&self) -> Result<Value> {
        get_value(self.conn, "api/net/net-mode")
    }

    /// Set the network mode and bands.
    ///
    /// * `lteband` — LTE band bitmask (`LTEBand::All` for all / non-4G).
    /// * `networkband` — 3G band bitmask (`NetworkBand::All` for all / non-3G).
    /// * `networkmode` — desired [`NetworkMode`].
    pub fn set_net_mode(
        &self,
        lteband: u64,
        networkband: u64,
        networkmode: NetworkMode,
    ) -> Result<String> {
        let body: XmlMap = map_of([
            ("NetworkMode", networkmode.as_str().to_string()),
            ("NetworkBand", format!("{networkband:x}")),
            ("LTEBand", format!("{lteband:x}")),
        ]);
        post_set(self.conn, "api/net/net-mode", &body)
    }

    /// `net/network`. Network information.
    pub fn network(&self) -> Result<Value> {
        get_value(self.conn, "api/net/network")
    }

    /// Set the network mode and band (different value ranges than
    /// [`Net::set_net_mode`]).
    pub fn set_network(&self, networkmode: &str, networkband: &str) -> Result<String> {
        let body: XmlMap = map_of([
            ("NetworkMode", networkmode.to_string()),
            ("NetworkBand", networkband.to_string()),
        ]);
        post_set(self.conn, "api/net/network", &body)
    }

    /// `net/register`. Network registration status.
    pub fn register(&self) -> Result<Value> {
        get_value(self.conn, "api/net/register")
    }

    /// Set the network registration (manual/auto selection).
    ///
    /// * `mode` — `"1"` manual, `"0"` auto.
    /// * `plmn` — PLMN code, `""` for auto.
    /// * `rat` — `"0"` 2G, `"2"` 3G, `"7"` 4G; `""` for auto.
    pub fn set_register(&self, mode: &str, plmn: &str, rat: &str) -> Result<String> {
        let body: XmlMap = map_of([
            ("Mode", mode.to_string()),
            ("Plmn", plmn.to_string()),
            ("Rat", rat.to_string()),
        ]);
        post_set(self.conn, "api/net/register", &body)
    }

    /// `net/net-mode-list`. Available network modes.
    pub fn net_mode_list(&self) -> Result<Value> {
        get_value(self.conn, "api/net/net-mode-list")
    }

    /// `net/plmn-list`. Available PLMNs.
    pub fn plmn_list(&self) -> Result<Value> {
        get_value(self.conn, "api/net/plmn-list")
    }

    /// `net/net-feature-switch`. Network feature switch status.
    pub fn net_feature_switch(&self) -> Result<Value> {
        get_value(self.conn, "api/net/net-feature-switch")
    }

    /// `net/cell-info`. Cell information.
    pub fn cell_info(&self) -> Result<Value> {
        get_value(self.conn, "api/net/cell-info")
    }

    /// `net/csps_state`. CSPS state.
    pub fn csps_state(&self) -> Result<Value> {
        get_value(self.conn, "api/net/csps_state")
    }

    /// Reconnect to the network.
    pub fn reconnect(&self) -> Result<String> {
        let body: XmlMap = map_of([("ReconnectAction", "1".to_string())]);
        post_set(self.conn, "api/net/reconnect", &body)
    }

    /// `net/antenna-configuration`. 5G antenna configuration.
    ///
    /// Supplementary endpoint from
    /// Brovi-Huawei-5G-CPE-Manager.
    pub fn get_antenna_configuration(&self) -> Result<Value> {
        get_value(self.conn, "api/net/antenna-configuration")
    }

    /// `net/lock-cell`. Lock the device to a specific 5G cell.
    ///
    /// * `lock` — `true` to lock, `false` to unlock.
    /// * `freq` — the cell frequency (ARFCN), `0` when unlocking.
    /// * `pci` — the physical cell identity, `0` when unlocking.
    ///
    /// Supplementary endpoint from
    /// Brovi-Huawei-5G-CPE-Manager.
    pub fn set_lock_cell(&self, lock: bool, freq: i64, pci: i64) -> Result<String> {
        let body: XmlMap = map_of([
            (
                "LockCell",
                if lock {
                    "1".to_string()
                } else {
                    "0".to_string()
                },
            ),
            ("Freq", freq.to_string()),
            ("PCI", pci.to_string()),
        ]);
        post_set(self.conn, "api/net/lock-cell", &body)
    }
}
